operational_analysis/methods/turbine_long_term_gross_energy_uq.py

```diff
@@ -0,0 +1,535 @@
+# gross energy. The main output is a 'waterfall' plot linking the EYA-
+# estimated and operational-estiamted AEP values. 
+
+import pandas as pd
+import numpy as np
+from tqdm import tqdm
+import random
+
+from operational_analysis.toolkits import met_data_processing
+from operational_analysis.toolkits import filters
+from operational_analysis.toolkits.power_curve import functions
+from operational_analysis.toolkits import imputing
+from operational_analysis.toolkits import timeseries
+
+from operational_analysis import logged_method_call
+from operational_analysis import logging
+
+logger = logging.getLogger(__name__)
+
+class TurbineLongTermGrossEnergy(object):
+    """
+    A serial (Pandas-driven) implementation of calculating long-term gross energy 
+    for each turbine in a wind farm. This module collects standard processing and
+    analysis methods for estimating this metric.
+    The method proceeds as follows:
+        1. Filter turbine data for normal operation
+        2. Calculate daily means of wind speed, wind direction, and air density from reanalysis products
+        3. Calculate daily sums of energy from each turbine
+        4. Fit daily data (features are atmospheric variables, response is turbine power) using a
+           generalized additive model (GAM)
+        5. Apply model results to long-term atmospheric varaibles to calculate long term
+           gross energy for each turbine
+    A Monte Carlo approach is implemented to repeat the procedure multiple times
+    to get a distribution of results, from which deriving uncertainty quantification
+    for the long-term gross energy estimate.
+    The end result is a table of long-term gross energy values for each turbine in the wind farm. Note
+    that this gross energy metric does not back out losses associated with waking or turbine performance.
+    Rather, gross energy in this context is what turbine would have produced under normal operation 
+    (i.e. excluding downtime and underperformance).
+    Required schema of PlantData:
+            - _scada_freq
+            - reanalysis products ['merra2', 'erai', 'ncep2'] with columns ['time', 'u_ms', 'v_ms', 'windspeed_ms', 'rho_kgm-3']
+            - scada with columns: ['time', 'id', 'wmet_wdspd_avg', 'wtur_W_avg', 'energy_kwh']
+    """
+
+    @logged_method_call
+    def __init__(self, plant, num_sim = 100, uncertainty_meter=0.005, uncertainty_scada=0.005,
+                  uncertainty_wind_bin_thresh=(1, 3), uncertainty_max_power_filter=(0.8, 0.9), 
+                  uncertainty_correction_threshold=(0.85, 0.95)):
+    
+        """
+        Initialize turbine long-term gross energy analysis with data and parameters.
+        Args:
+         plant(:obj:`PlantData object`): PlantData object from which TurbineLongTermGrossEnergy should draw data.
+         num_sim:(:obj:`int`): number of Monte Carlo simulations
+         uncertainty_meter(:obj:`float`): uncertainty imposed to revenue meter data
```

Comment:
uncertainty_meter isn't used in this class, right? So it should be removed from the arguments.

Reply:
Right thanks, I will remove that!

Reply:
Has this been resolved?

Reply:
Yes, I've now marked this as resolved.